### PR TITLE
Mark "run once" upgrades as run on fresh install

### PR DESCRIPTION
**Pull request by seadowg**
Work towards #4927 (we'll need to do another PR to fix on `master`)

This can be a little hard to get your head around. Basically, we have a way of saying upgrades should only run once, but we never considered the fact that we also didn't want those upgrades to ever run if they exist during a fresh install. This changes our upgrade process so that we mark "run once" upgrades (currently just the existing project migration) as done when doing a fresh install.

#### What has been done to verify that this works as intended?

New tests.

#### Why is this the best possible solution? Were any other approaches considered?

I also considered cleaning up the extra project created due to the bug but couldn't find a way to identify it safely.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Should just fix the bug! Good to check upgrades (from different versions) and fresh installs. As @grzesiek2010 pointed out, installing a previous version and then this version from source will not simulate an upgrade as the version code used will be the same (it's based on the number of commits on the `master` branch). To simulate an upgrade:

1. Check out the last released version (`v2021.3.2`)
2. Set `versionCode` in `collect_app/build.gradle` to `1`
3. Install on device (make sure to uninstall first)
4. Check out the PR branch again
5. Set `versionCode` in `collect_app/build.gradle` to `2`
6. Install on device

This simulates a "real" upgrade where the version code would change as well as the code.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by lognaturel** [comment]
I will forever be shocked at what users don’t report. Glad you found it, @kkrawczyk123! Looks good to me so I’ll mark as needing testing for @getodk/testers to start right away. Would still be good for you to take a carefully look, @grzesiek2010. 🙏

**Comment by yanokwa** [comment]
I just got a report about this from an ODK Cloud user a few hours ago. So they do report things, but @kkrawczyk123 is a lot faster and more reliable!

**Comment by grzesiek2010** [comment]
I wanted to dig deeper to fully understand what is happening but I can't seem to reproduce the issue. I tried on two emulators and one real device those steps:
1. Install v2021.3.1
2. Select Demo project
3. Install v2021.3.2
4. And I have just on project

@seadowg did you have to do anything else? I talked to @kkrawczyk123 but she claims that those steps are enough in her case.

**Comment by srujner** [comment]
@grzesiek2010  I was able to reproduce the issue using the steps provided by @kkrawczyk123 on Android 10 and 11.

**Comment by seadowg** [comment]
@grzesiek2010 those are the steps I followed. An extra subtlety is that you need to do a fresh install of v2021.3.1 - although you should be able to reproduce with any version from v2021.2 onwards really. What's happening in terms of the repro steps is:

1. Install v2021.3.1 - app starts up and does not run `AppUpgrader.upgrade` as we're doing a fresh install
2. Select Demo project - demo project is setup with
3. Install v2021.3.2 - app starts up and runs `AppUpgrader.upgrade` which ends up running `ExistingProjectMigrator` as it was never run before

We end up with a project called "Demo project" because the legacy general and admin settings are empty so `ProjectDetailsCreator` gives us that project name due to us using the default URL (see `ProjectDetailsCreator#getProjectNameFromConnectionIdentifier`).

**Comment by grzesiek2010** [comment]
Ok I was finally able to reproduce the issue...

File: collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java]: Doesn't it mean that if you have v2021.3 and install v2021.4 (assuming we are going to release it with this fix) it will still go and perform `appUpgrader.upgrade()` first and run those migrations?

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java]: I'm pretty sure I'm right...

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java]: Errr maybe? Just going to try and write this out to get it straight in my head:

* Fresh install of 2021.3.2 (or any version with projects) and upgrade to 2022.1: `upgrade()` runs without ever running `install()` so they get an extra project 👎
* Fresh install of 2021.3.3 (next hot fix) and upgrade to 2022.1: `upgrade()` runs but after `install()` on fresh install 👍
* Upgrade to 2021.3.3 from 2021.3.2 with previous upgrade from 2021.3.1 (or any version with projects): `upgrade()` runs but after `ExistingProjectMIgrator` has already run on previous upgrades 👍

So yeah I think you're correct that this hasn't quite fixed the problem. If we'd included `install()` in our original release with `ExistingProjectMigrator` everything would work but as you point out, we need to account for the existing bad state. 

I think to correct the problem we need to run `install()` before `upgrade`() when upgrading from a version before 2021.3.3 (our next hot fix) and after 1.30.1 (all versions after this have projects and `ExistingProjectMigrator`). Does that sound right? We'll need to work out version codes for these as well.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java]: Can't you just base your fix on `LAST_LAUNCHED` value that we keep in meta settings? If that value represents a version that already had projects the migration should be skipped.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java]: Yeah sorry I should have worded that differently, calling `install()` is a bit over the top. How about: "stop `ExistingProjectMigrator` from running when upgrading from a version after 1.30.1 (using version code to determine the last version)". Or actually, even simpler: "stop `ExistingProjectMigrator` running when upgrading from a version without `LAST_LAUNCHED`" as `LAST_LAUNCHED` was introduced as part of 2021.2.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java]: If `LAST_LAUNCHED`  was released in 2021.2 it won't be a perfect fix because still upgrading from 2021.0 and 2021.1 will be broken but that doesn't seem to be a big deal. Alternatively you can think of another solution like for example:
`If projects repository is not empty that means we do upgrade from a version with projects and can skip migrations.`

**Comment by grzesiek2010** [comment]
Another thing: I'm pretty sure the QA team is not able to verify this fix correctly because the flow requires installing an apk with a newer version on an existing one. If we build apks on our own no matter from which commit we start the version will be always the same because it's calculated in gradle:
`LEGACY_BUILD_NUMBER_OFFSET + getMasterCommitCount()`
where `LEGACY_BUILD_NUMBER_OFFSET` is a static number 1067 and `getMasterCommitCount()` it's a git command
`git rev-list --first-parent --count master` which will return the same number no matter on what branch you are because it takes master and your master branch is always in the same state.
That means we always end up with the same version number and upgrading won't take place at all.
To reproduce the issue you used released apks and there  those versions differ of course but we can't install our builds on release builds.

So to be able to verify the fix correctly the QA team would need an additional commit with bumped version number. Am I right?

File: collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java]: There's no 2021.0 or .1. 1.30 was the last version before projects and 2021.2 first version with them. We **want** upgrading from 1.30 (and before) to run `ExistingProjectMigrator` and there is already the upgrade `key` concept that will take care of making sure that the upgrade doesn't run again after that, so I think we're fine on older versions.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.java]: > There's no 2021.0

Right my mistake then ok it will be fine.

**Comment by seadowg** [comment]
> Another thing: I'm pretty sure the QA team is not able to verify this fix correctly because the flow requires installing an apk with a newer version on an existing one.

@grzesiek2010 great point. I'm taking `needs testing` off this right now anyway as we're pretty sure the fix isn't good enough (cc/ @sruner and @kkrawczyk123). You're right that to simulate a true upgrade we're going to need to change `versionCode` in `collect_app`/`build.gradle`. I'll work on improving this fix based on my conversation with @grzesiek2010 and also post a workflow for simulating the upgrade.

**Comment by seadowg** [comment]
Ok that's hopefully in a better place now. I've also included steps to simulate an upgrade in the PR description. Will wait for approval to mark this ready for testing.

File: collect_app/src/main/java/org/odk/collect/android/application/initialization/upgrade/AppUpgrader.kt
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/application/initialization/upgrade/AppUpgrader.kt]: I'm not sure if I understand the purpose of keeping it... if you keep that and run in v2021.3.3 that is going to set flags for all migrators to true (do not perform in the future). Maybe not in the first run when you upgrade from an older version but then if you kill the app and open again it will happen.

Let's say we have a user who uses v2021.3.3 so all those flags are set to true on his device.
Then in v2022.1 you decide to rename one of our preferences (`ExistingSettingsMigrator` is responsible for that).
Our user upgrades form v2021.3.3 to v2022.1 but his `ExistingSettingsMigrator` is not trigger because the flag of this migrator was set to true.

Am i missing something?

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/upgrade/AppUpgrader.kt]: The thing is that we have two types of migrators:
- final migartors
- editable migrators

`ExistingProjectMigrator` is final if it was performed once we shouldn't need it anymore so setting it's flag to true is fine.
`ExistingSettingsMigrator` I would say it's not final we might need it every time we decide to update some prefs
`FormUpdatesUpgrade` here I'm not sure I would need to dig depper.

so you can't disable all of them unless I'm missing something.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/upgrade/AppUpgrader.kt]: `Upgrades` can only be skipped if they have a non `null` `key()` (you can see in `AppUpgrader` or its tests). This lets us run `ExistingSettingsMigrator` and `FormUpdatesUpgrade` every upgrade and only run `ExistingProjectMigrator` once. Calling `install()` will not cause the first two to be skipped as they return `null` from their `key()` implementations so can't be marked as already having been run. We currently don't **need** the call to `install()` because we've put a check in `ExistingProjectMigrator` but it lets us add another "run once" (or final as you say) upgrade. Basically, the upgrade logic wasn't correct before but now is in my opinion.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/upgrade/AppUpgrader.kt]: Ok now I see but the way it woks (with those keys null or nonnull) is confusing...
So this will be run only for `ExistingProjectMigrator` so why is it important if we do the check inside `ExistingProjectMigrator` either way?

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/initialization/upgrade/AppUpgrader.kt]: Ok for this PR we can just take out the `install()` call as I think it's confusing things. I do want this to go in `master` but we can discuss that separately as improvements to the way upgrades work rather than just a fix for this specific issue.

**Review by grzesiek2010 — APPROVED** [review]

**Comment by lognaturel** [comment]
Thanks for really thinking through it, @grzesiek2010!

**Comment by srujner** [comment]
**Tested with success!**
Verified on Androids: 10, 11

Verified cases:
- Update with only demo project created
- Update with multiple projects created
- Update without demo project 
- Update without any projects added
- Changed settings before updating the app
- Adding a project with qr code and then updating the app
- Update from .apk and Android studio

**Comment by kkrawczyk123** [comment]
Verified also on Androids: 5.1, 8.1 and 9.

**Pull request merged**
